pkg/types/query/query.go

```diff
@@ -0,0 +1,294 @@
+func Comparer(name string, valueComparers ...ValueComparer) Expression {
+	return Expression{Primitive: &ComparerPrimitive{Name: name, ValueComparers: valueComparers}}
+}
+
+func (f *ComparerPrimitive) Accept(visitor Visitor) {
+	visitor.ComparerPrimitive(*f)
+}
+
+// BlockPrimitive is a primitive of KeyFilter that filters search in comparison to block number.
+type BlockPrimitive struct {
+	Block    uint64
+	Operator ComparisonOperator
+}
+
+func Block(block uint64, operator ComparisonOperator) Expression {
+	return Expression{
+		Primitive: &BlockPrimitive{Block: block, Operator: operator},
+	}
+}
+
+func (f *BlockPrimitive) Accept(visitor Visitor) {
+	visitor.BlockPrimitive(*f)
+}
+
+type ConfirmationLevel int32
+
+const (
+	Finalized   = ConfirmationLevel(0)
+	Unconfirmed = ConfirmationLevel(1)
+)
+
+// ConfirmationsPrimitive is a primitive of KeyFilter that filters search to results that have a certain level of confirmation.
+// Confirmation map to different concepts on different blockchains.
+type ConfirmationsPrimitive struct {
+	ConfirmationLevel
+}
+
+func Confirmation(confLevel ConfirmationLevel) Expression {
+	return Expression{
+		Primitive: &ConfirmationsPrimitive{ConfirmationLevel: confLevel},
+	}
+}
+
+func (f *ConfirmationsPrimitive) Accept(visitor Visitor) {
+	visitor.ConfirmationPrimitive(*f)
+}
+
+// TimestampPrimitive is a primitive of KeyFilter that filters search in comparison to timestamp.
+type TimestampPrimitive struct {
+	Timestamp uint64
+	Operator  ComparisonOperator
+}
+
+func Timestamp(timestamp uint64, operator ComparisonOperator) Expression {
+	return Expression{
+		Primitive: &TimestampPrimitive{timestamp, operator},
+	}
+}
+
+func (f *TimestampPrimitive) Accept(visitor Visitor) {
+	visitor.TimestampPrimitive(*f)
+}
+
+// TxHashPrimitive is a primitive of KeyFilter that filters search to results that contain txHash.
+type TxHashPrimitive struct {
+	TxHash string
+}
+
+func TxHash(txHash string) Expression {
+	return Expression{
+		Primitive: &TxHashPrimitive{txHash},
+	}
+}
+
+func (f *TxHashPrimitive) Accept(visitor Visitor) {
+	visitor.TxHashPrimitives(*f)
+}
+
+// Where is a helper function for building KeyFilter, eg. usage:
+//
+//	 queryFilter, err := Where(
+//
+//				TxHash("0xHash"),
+//				And(Block(startBlock, Gte),
+//					Block(endBlock, Lte)),
+//					Or(
+//						And(
+//							Timestamp(someTs1, Gte),
+//							Timestamp(otherTs1, Lte)),
+//						And(
+//							Timestamp(someTs2, Gte),
+//							Timestamp(otherTs2, Lte))
+//					)
+//			  	 )
+//		 ==> `txHash = txHash AND (
+//									 block > startBlock AND block < endBlock
+//									 AND (
+//										 (timestamp > someTs1 And timestamp < otherTs1)
+//										 OR
+//										 (timestamp > someTs2 And timestamp < otherTs2)
+//									 )
+//							 )`
+//		if err != nil{return nil, err}
+//		QueryKey(key, queryFilter)...
```

Comment:
Formatted it better now. Examples in tests will exist, but will be on the relayer side. I thought it would be nice to have an example on hover since it may be confusing at first glance